vendor/brave-ios/Ledger/Data/BATLedgerDatabase.mm

```diff
@@ -627,4 +641,245 @@ + (double)reservedAmountForPendingContributions
+      NSError *error = nil;
+      const auto deleteRequest = [[NSBatchDeleteRequest alloc] initWithFetchRequest:fetchRequest];
+      [context executeRequest:deleteRequest error:&error];
+      
+      if (context.hasChanges) {
+        [context save:nil];
+      }
+    }];
+    BLOG(ledger::LogLevel::LOG_INFO) << "CoreData: Deleted publisher list in " << [[NSDate date] timeIntervalSinceDate:start] << " seconds" << std::endl;
+    const auto operationQueue = [[NSOperationQueue alloc] init];
+    // 2 concurrent ops operate better than 4 or 6 when testing
+    operationQueue.maxConcurrentOperationCount = MIN(2, [[NSProcessInfo processInfo] activeProcessorCount]);
+    operationQueue.name = @"Publisher List Insertions";
+    BLOG(ledger::LogLevel::LOG_INFO) << "CoreData: Inserting " << list.count << " publisher info objects" << std::endl;
+    for (NSArray *section in split) {
+      [operationQueue addOperationWithBlock:^{
+        const auto newContext = [DataController newBackgroundContext];
+        [newContext performBlockAndWait:^{
+          [self insertOrUpdateServerPublisherList:section context:newContext];
+          if (newContext.hasChanges) {
+            NSError *error;
+            if (![newContext save:&error]) {
+              BLOG(ledger::LogLevel::LOG_ERROR) << "CoreData: Save error: " << error.debugDescription << std::endl;
+            }
+          }
+        }];
+      }];
+    }
+    [operationQueue waitUntilAllOperationsAreFinished];
+    NSTimeInterval duration = [[NSDate date] timeIntervalSinceDate:start];
+    BLOG(ledger::LogLevel::LOG_INFO) << "CoreData: Completed publisher list task. Took total of " << duration << " seconds" << std::endl;
+    dispatch_async(dispatch_get_main_queue(), ^{
+      completion(true);
+    });
+  });
+}
+
+#pragma mark - Publisher List / Banner
+
++ (nullable BATPublisherBanner *)bannerForPublisherID:(NSString *)publisherID
+{
+  const auto spb = [self getServerPublisherBannerWithPublisherID:publisherID context:DataController.viewContext];
+  if (!spb) {
+    return nil;
+  }
+  const auto banner = [[BATPublisherBanner alloc] init];
+  banner.publisherKey = publisherID;
+  banner.title = spb.title;
+  banner.desc = spb.desc;
+  banner.background = spb.background;
+  banner.logo = spb.logo;
+  banner.links = [self publisherLinksWithPublisherID:publisherID];
+  banner.amounts = [self bannerAmountsForPublisherWithPublisherID:publisherID];
+  return banner;
+}
+
++ (void)insertOrUpdateBannerForServerPublisherInfo:(ServerPublisherInfo *)info
+                                            banner:(BATPublisherBanner *)banner
+                                           context:(NSManagedObjectContext *)context
+{
+  if (!banner) {
+    return;
+  }
+  
+  auto spb = [[ServerPublisherBanner alloc] initWithEntity:[NSEntityDescription entityForName:NSStringFromClass(ServerPublisherBanner.class) inManagedObjectContext:context] insertIntoManagedObjectContext:context];
+  spb.serverPublisherInfo = info;
```

Comment:
See my other comment about context mismatching.

I see that when this method is called we create a new object using `context`, and then pass the same `context`

hmm perhaps we could skip the `context` argument and pull it directly out of the object by using `info.managedObjectContext`

Reply:
Is that safe to do? I know for sure that the context I pass in is non-nil but the doc on `managedObjectContext` is nullable ("May be nil if the receiver has been deleted from its context")

Reply:
I agree, I would take the safer route and just compare two contexts